### ISSUE TITLE
[Feature Request] Debug facilities

### ISSUE THREAD

**Issue by jm-begon** [montefiore-ai/alan-cluster#8]
Some errors cannot be debugged locally on CPU (typical example: mismatch of CPU and GPU tensors in PyTorch). It would be nice to have an environment as close as the one in which the code will eventually run for that purpose.

There are four approches I can think of
 - Use the usual queue to submit job with low time requirement.
 - Use a debug queue with high priority and low maximum time bound.
 - Use a debug queue with a dedicated GPU.
 - Leave one GPU out of the computing nodes altogether.

The first approach does not require any change but offers no guarantee. Also, I am not sure how the fairness measures will be influenced.

The second is more principled but there is still no guarantee that a code can de debugged quickly if all GPUs are busy. Usually, one would want to debug the code in real time.

Finally leaving one GPU out (either explicitly or through a dedicated queue) allows for instantaneous debugging even when the cluster is under heavy load. On the other hand, it also implies that one GPU is rarely used.

I guess the third solution is best if we are ready to sacrifice one GPU.

@waliens @AWehenkel

**Comment by glouppe** [comment]
At VSCentrum (the Flemish cluster), they have 'debug nodes' that are set aside (apparently) for this purpose. https://www.vscentrum.be/cluster-doc/running-jobs/job-system-basics

I am not sure we can afford to lose one or more GPUs though. At least not yet.

**Comment by waliens** [comment]
@JoeriHermans As alan has now 50+ gpus, do you think we might set aside one or two for debugging at this point ?

**Comment by JoeriHermans** [comment]
@jm-begon @waliens @glouppe I can prep 2 GPU's in alan-compute-05 for the debug partition. What time-limit do you think is appropriate? I was thinking ~15 min max.

**Comment by JoeriHermans** [comment]
@AWehenkel @jm-begon @waliens @glouppe 

When updating the configuration to account for the debug partition I noticed I cannot isolate separate GPU's in the debug partition, only distinct nodes. So you are OK with using `alan-compute-05` as a debug node (as it only has 64GB of RAM). It is a waste though imho, but if you really desire this I can configure the cluster as such.

**Comment by waliens** [comment]
CPU testing should work indeed, however I am not sure about the time it would take.

**Comment by JoeriHermans** [comment]
I mean, the purpose is to test the code. So like a few iterations of batches should be fine right? If you check if your model makes sense, you'll just have to train it.

**Comment by jm-begon** [comment]
Usually, you would want to at least run for one epoch. I guess we can start with 15 minutes (for GPU) and if it happens to be really too short we can discuss it further. On CPU 15 minutes is really too short. 

People should indeed be encouraged to test code on CPU first but--as you mention it--this would not catch device problems. Being able to troubleshoot on GPU is a very nice feature. However, I do agree that losing 4 GPUs for that might be a lot. If it is not too much trouble, I propose we test it that way and evaluate the matter in some near future.

At some point, someone might get some fundings which are insufficient for a whole top-notch, 4-GPU node but enough for a moderate, 2-GPU  one.

**Comment by JoeriHermans** [comment]
> this would not catch device problems.

Could you give an example? I can only come up with memory issues and resource utilization. From what I've seen, the latter is typically ignored because nobody actually optimizes the code to maximize the GPU utilization. 

To be honest, I think we do not really require this because once the new scheduler is installed we will have more "fair" partitioning of resources anyway.

**Comment by waliens** [comment]
Yesterday, I had to investigate an issue that occurred only after few epochs. On GPU, it took 7 minutes but on CPU, the 1 hour time limit was not enough to reach this point. To be completely fair, I should say I have only used 1 CPU, using at least two should already improve duration. Still, I am not sure I would gain that much, and even 7 minutes is already long when you want to see the effect of a fix. 

Anyway, my code finally works and I don't need to debug anything right now as far as I am concerned. So, if no one needs to debug code, maybe we can postpone this until someone buys a small debug node as suggested by @jm-begon or the situation solves itself with the new scheduler.

**Comment by jm-begon** [comment]
@JoeriHermans, a recurring bug is to have some portion of code lying on CPU and the other on GPU, when you forget to `.to(device)` something. 

If you believe the new scheduling policy will help, we can re-assess the matter later.

Also, if GPU utilization is a problem, I think we should spend some time discussing it all together.

@waliens, if you have to change your code to make it work on several CPU just to debug, I don't think this is a good approach.
